nilearn/mass_univariate/permuted_least_squares.py

```diff
@@ -446,92 +652,234 @@ def permuted_ols(
-                                    * sign_scores_original_data)
         return np.asarray([]), scores_original_data.T, np.asarray([])
 
     # actual permutations, seeded from a random integer between 0 and maximum
     # value represented by np.int32 (to have a large entropy).
     ret = joblib.Parallel(n_jobs=n_jobs, verbose=verbose)(
         joblib.delayed(_permuted_ols_on_chunk)(
-            scores_original_data, testedvars_resid_covars,
-            targetvars_resid_covars.T, thread_id + 1, covars_orthonormalized,
-            n_perm=n_perm, n_perm_chunk=n_perm_chunk,
-            intercept_test=intercept_test, two_sided_test=two_sided_test,
+            scores_original_data,
+            testedvars_resid_covars,
+            targetvars_resid_covars.T,
+            thread_id=thread_id + 1,
+            threshold=threshold_t,
+            confounding_vars=covars_orthonormalized,
+            masker=masker,
+            n_perm=n_perm,
+            n_perm_chunk=n_perm_chunk,
+            intercept_test=intercept_test,
+            two_sided_test=two_sided_test,
             random_state=rng.randint(1, np.iinfo(np.int32).max - 1),
-            verbose=verbose)
+            verbose=verbose,
+        )
         for thread_id, n_perm_chunk in enumerate(n_perm_chunks))
-    # reduce results
-    scores_as_ranks_parts, h0_fmax_parts = zip(*ret)
-    h0_fmax = np.hstack((h0_fmax_parts))
-    scores_as_ranks = np.zeros((n_regressors, n_descriptors))
-    for scores_as_ranks_part in scores_as_ranks_parts:
-        scores_as_ranks += scores_as_ranks_part
-    # convert ranks into p-values
-    pvals = (n_perm + 1 - scores_as_ranks) / float(1 + n_perm)
-
-    # put back sign on scores if it was removed in the case of a two-sided test
-    # (useful to distinguish between positive and negative effects)
-    if two_sided_test:
-        scores_original_data = scores_original_data * sign_scores_original_data
 
-    return - np.log10(pvals), scores_original_data.T, h0_fmax
+    # reduce results
+    (
+        vfwe_scores_as_ranks_parts,
+        vfwe_h0_parts,
+        csfwe_h0_parts,
+        cmfwe_h0_parts,
+    ) = zip(*ret)
+
+    # Voxel-level FWE
+    vfwe_h0 = np.hstack((vfwe_h0_parts))
+    vfwe_scores_as_ranks = np.zeros((n_regressors, n_descriptors))
+    for vfwe_scores_as_ranks_part in vfwe_scores_as_ranks_parts:
+        vfwe_scores_as_ranks += vfwe_scores_as_ranks_part
+
+    vfwe_pvals = (n_perm + 1 - vfwe_scores_as_ranks) / float(1 + n_perm)
+
+    if threshold is not None:
+        # Cluster-size and cluster-mass FWE
+        csfwe_h0 = np.hstack((csfwe_h0_parts))
+        cmfwe_h0 = np.hstack((cmfwe_h0_parts))
+
+        csfwe_pvals = np.zeros_like(vfwe_pvals)
+        cmfwe_pvals = np.zeros_like(vfwe_pvals)
+
+        # TODO: Eliminate need to transpose
+        scores_original_data_4d = masker.inverse_transform(
+            scores_original_data.T
+        ).get_fdata()
+        conn = ndimage.generate_binary_structure(3, 1)
```

Comment:
I think that in other parts of the library , we use a different définition for `conn`.

Reply:
:+1: I'll change it to `bin_struct` (for binary structure).

Reply:
Sorry, what I meant is that the specification of the structure is different e.g. in 
https://github.com/nilearn/nilearn/blob/0e678341aee2866527d99fb9344cfa5ae56aa7e3/nilearn/image/image.py#L793
I think we should strive to have only one definition across the library and always use it.

Reply:
ohhhhh okay. I'll change the other cases then.